File: clients/go/termwright/tree.go
package termwright

import (
	"fmt"
	"strconv"
	"strings"

	"github.com/rivo/tview"

	"github.com/gorce-ai/termwright/clients/go/protocol"
)

// maxTextName bounds the accessible name taken from a free-text widget, so a
// scrolling log cannot inflate every snapshot.
const maxTextName = 200

// buildSnapshot reads the primitive tree into a snapshot. It only reads: no
// primitive is drawn, focused or resized by this walk.
func (s *Session) buildSnapshot(columns, rows int) *protocol.Snapshot {
	snapshot := protocol.NewSnapshot("", 0, columns, rows)
	s.walk(s.root, "", snapshot, columns, rows, false)
	return snapshot
}

func (s *Session) walk(
	primitive tview.Primitive,
	parentID string,
	snapshot *protocol.Snapshot,
	columns, rows int,
	hidden bool,
) {
	if primitive == nil {
		return
	}

	id := s.idFor(primitive)
	role, name := s.describe(primitive)
	children := s.childrenOf(primitive)

	// HasFocus is true for ancestors of the focused primitive as well, so the
	// focus flag belongs to the deepest primitive that reports it. Asking the
	// application instead would deadlock: this runs inside tview's draw lock.
	focused := primitive.HasFocus() && !anyHasFocus(children)
	if hidden {
		// Nothing off-screen holds the focus, whatever the primitive believes.
		focused = false
	}

	node := protocol.Node{
		ID:            id,
		ParentID:      parentID,
		Role:          role,
		Name:          name,
		Bounds:        boundsOf(primitive, columns, rows),
		State:         stateOf(primitive, focused, hidden),
		Actions:       actionsFor(role),
		Value:         valueOf(primitive),
		TestID:        s.testIDFor(primitive),
		FrameworkType: frameworkTypeFor(primitive, role),
	}
	if parentID == "" {
		snapshot.RootIDs = append(snapshot.RootIDs, id)
	}
	snapshot.Nodes = append(snapshot.Nodes, node)

	for _, child := range children {
		// Hiding is inherited: everything under an unshown page is unshown.
		s.walk(child.primitive, id, snapshot, columns, rows, hidden || child.hidden)
	}
	s.appendSynthetic(primitive, id, snapshot, hidden)
}

// idFor assigns a stable id per primitive, so a node keeps its identity across
// revisions and the driver can follow it.
func (s *Session) idFor(primitive tview.Primitive) string {
	s.mu.Lock()
	defer s.mu.Unlock()
	if id, ok := s.ids[primitive]; ok {
		return id
	}
	s.nextID++
	id := "p" + strconv.Itoa(s.nextID)
	s.ids[primitive] = id
	return id
}

// -- roles and names -------------------------------------------------------

func (s *Session) describe(primitive tview.Primitive) (protocol.Role, string) {
	if s.config.describe != nil {
		if role, name, ok := s.config.describe(primitive); ok {
			return role, name
		}
	}
	return roleOf(primitive), nameOf(primitive)
}

// roleOf maps a tview type onto a v1 semantic role.
func roleOf(primitive tview.Primitive) protocol.Role {
	switch primitive.(type) {
	case *tview.Button:
		return protocol.RoleButton
	case *tview.Checkbox:
		return protocol.RoleCheckbox
	case *tview.InputField, *tview.TextArea:
		return protocol.RoleTextbox
	case *tview.DropDown, *tview.List, *tview.TreeView:
		return protocol.RoleList
	case *tview.Table:
		return protocol.RoleTable
	case *tview.TextView:
		return protocol.RoleText
	case *tview.Modal:
		return protocol.RoleDialog
	case *tview.Form, *tview.Flex, *tview.Grid, *tview.Pages, *tview.Frame:
		return protocol.RoleRegion
	case *tview.Box:
		return protocol.RoleRegion
	}
	return protocol.RoleGeneric
}

// nameOf derives the accessible name: an explicit label wins, then the box
// title, then the widget's own text.
func nameOf(primitive tview.Primitive) string {
	switch widget := primitive.(type) {
	case *tview.Button:
		return widget.GetLabel()
	case *tview.Checkbox:
		return firstNonEmpty(widget.GetLabel(), widget.GetTitle())
	case *tview.InputField:
		return firstNonEmpty(widget.GetLabel(), widget.GetTitle())
	case *tview.DropDown:
		return firstNonEmpty(widget.GetLabel(), widget.GetTitle())
	case *tview.TextArea:
		return firstNonEmpty(widget.GetLabel(), widget.GetTitle())
	case *tview.TextView:
		return firstNonEmpty(widget.GetTitle(), trimText(widget.GetText(true)))
	case *tview.Box:
		return widget.GetTitle()
	}
	if boxed, ok := primitive.(interface{ GetTitle() string }); ok {
		return boxed.GetTitle()
	}
	return ""
}

// valueOf reports what a value-bearing widget contains, including the empty
// string — an empty field has a value of "", not no value. A nil result means
// the widget is not value-bearing at all.
func valueOf(primitive tview.Primitive) *string {
	switch widget := primitive.(type) {
	case *tview.InputField:
		return protocol.Text(widget.GetText())
	case *tview.TextArea:
		return protocol.Text(trimText(widget.GetText()))
	case *tview.DropDown:
		_, option := widget.GetCurrentOption()
		return protocol.Text(option)
	}
	return nil
}

func trimText(text string) string {
	text = strings.TrimSpace(text)
	if len(text) > maxTextName {
		return text[:maxTextName]
	}
	return text
}

func firstNonEmpty(candidates ...string) string {
	for _, candidate := range candidates {
		if trimmed := strings.TrimSpace(candidate); trimmed != "" {
			return trimmed
		}
	}
	return ""
}

// -- geometry and state ----------------------------------------------------

// boundsOf returns absolute viewport bounds, or nil when the primitive is not
// on screen. tview's GetRect is already absolute, so no parent offset applies.
func boundsOf(primitive tview.Primitive, columns, rows int) *protocol.Rect {
	x, y, width, height := primitive.GetRect()
	if width <= 0 || height <= 0 {
		return nil
	}
	if x >= columns || y >= rows || x+width <= 0 || y+height <= 0 {
		return nil
	}
	return &protocol.Rect{Row: y, Column: x, Width: width, Height: height}
}

// anyHasFocus reports whether the focus lives below this level.
func anyHasFocus(children []child) bool {
	for _, entry := range children {
		if entry.primitive != nil && entry.primitive.HasFocus() {
			return true
		}
	}
	return false
}

func stateOf(primitive tview.Primitive, focused bool, hidden bool) *protocol.State {
	state := protocol.State{}
	empty := true

	if focused {
		state.Focused = protocol.Bool(true)
		empty = false
	}
	if hidden {
		state.Hidden = protocol.Bool(true)
		empty = false
	}
	switch widget := primitive.(type) {
	case *tview.Button:
		if widget.IsDisabled() {
			state.Disabled = protocol.Bool(true)
			empty = false
		}
	case *tview.Checkbox:
		state.Checked = widget.IsChecked()
		empty = false
	case *tview.Modal:
		state.Modal = protocol.Bool(true)
		empty = false
	case *tview.List:
		state.SetSize = protocol.Int(widget.GetItemCount())
		empty = false
	case *tview.Table:
		state.SetSize = protocol.Int(widget.GetRowCount())
		empty = false
	}
	if empty {
		return nil
	}
	return &state
}

func actionsFor(role protocol.Role) []protocol.Action {
	switch role {
	case protocol.RoleButton, protocol.RoleMenuItem, protocol.RoleTab:
		return []protocol.Action{protocol.ActionFocus, protocol.ActionActivate}
	case protocol.RoleCheckbox, protocol.RoleRadio:
		return []protocol.Action{protocol.ActionFocus, protocol.ActionToggle}
	case protocol.RoleTextbox:
		return []protocol.Action{protocol.ActionFocus, protocol.ActionSetValue}
	case protocol.RoleList, protocol.RoleTable:
		return []protocol.Action{protocol.ActionFocus, protocol.ActionScroll, protocol.ActionSelect}
	case protocol.RoleListItem:
		return []protocol.Action{protocol.ActionSelect, protocol.ActionActivate}
	}
	return nil
}

// -- child enumeration -----------------------------------------------------

// child is one enumerated child and whether its container is showing it.
type child struct {
	primitive tview.Primitive
	hidden    bool
}

func visible(primitives ...tview.Primitive) []child {
	children := make([]child, 0, len(primitives))
	for _, primitive := range primitives {
		if primitive != nil {
			children = append(children, child{primitive: primitive})
		}
	}
	return children
}

// childrenOf walks the containers tview exposes accessors for. Grid has no
// item accessor, so its children are reachable only through WithChildren.
//
// Pages is the one container that keeps children it is not showing, and those
// have to be published as hidden: a test asserting "the settings screen is
// open" must not pass while that page is still stacked out of sight.
func (s *Session) childrenOf(primitive tview.Primitive) []child {
	if s.config.children != nil {
		if supplied := s.config.children(primitive); supplied != nil {
			// A caller-supplied list says what exists, not what is shown; the
			// container's own state still decides visibility.
			return visible(supplied...)
		}
	}

	switch container := primitive.(type) {
	case *tview.Flex:
		primitives := make([]tview.Primitive, 0, container.GetItemCount())
		for index := 0; index < container.GetItemCount(); index++ {
			primitives = append(primitives, container.GetItem(index))
		}
		return visible(primitives...)
	case *tview.Pages:
		shown := make(map[string]struct{})
		for _, name := range container.GetPageNames(true) {
			shown[name] = struct{}{}
		}
		names := container.GetPageNames(false)
		children := make([]child, 0, len(names))
		for _, name := range names {
			page := container.GetPage(name)
			if page == nil {
				continue
			}
			_, isShown := shown[name]
			children = append(children, child{primitive: page, hidden: !isShown})
		}
		return children
	case *tview.Form:
		primitives := make([]tview.Primitive, 0, container.GetFormItemCount()+container.GetButtonCount())
		for index := 0; index < container.GetFormItemCount(); index++ {
			primitives = append(primitives, container.GetFormItem(index))
		}
		for index := 0; index < container.GetButtonCount(); index++ {
			primitives = append(primitives, container.GetButton(index))
		}
		return visible(primitives...)
	case *tview.Frame:
		if inner := container.GetPrimitive(); inner != nil {
			return visible(inner)
		}
	}
	return nil
}

// appendSynthetic emits nodes for items that are not primitives of their own —
// list entries and dropdown options — so they are addressable by role and name.
// They carry no bounds, which the schema allows.
func (s *Session) appendSynthetic(primitive tview.Primitive, parentID string, snapshot *protocol.Snapshot, hidden bool) {
	hiddenFlag := func() *bool {
		if hidden {
			return protocol.Bool(true)
		}
		return nil
	}
	switch widget := primitive.(type) {
	case *tview.List:
		current := widget.GetCurrentItem()
		for index := 0; index < widget.GetItemCount(); index++ {
			main, secondary := widget.GetItemText(index)
			node := protocol.Node{
				ID:       parentID + ":item" + strconv.Itoa(index),
				ParentID: parentID,
				Role:     protocol.RoleListItem,
				Name:     firstNonEmpty(main, secondary),
				Actions:  actionsFor(protocol.RoleListItem),
				State: &protocol.State{
					Selected:      protocol.Bool(index == current),
					PositionInSet: protocol.Int(index + 1),
					SetSize:       protocol.Int(widget.GetItemCount()),
					Hidden:        hiddenFlag(),
				},
			}
			snapshot.Nodes = append(snapshot.Nodes, node)
		}
	case *tview.DropDown:
		current, _ := widget.GetCurrentOption()
		for index := 0; index < widget.GetOptionCount(); index++ {
			snapshot.Nodes = append(snapshot.Nodes, protocol.Node{
				ID:       parentID + ":option" + strconv.Itoa(index),
				ParentID: parentID,
				Role:     protocol.RoleListItem,
				Name:     optionText(widget, index),
				Actions:  actionsFor(protocol.RoleListItem),
				State: &protocol.State{
					Selected:      protocol.Bool(index == current),
					PositionInSet: protocol.Int(index + 1),
					SetSize:       protocol.Int(widget.GetOptionCount()),
					Hidden:        hiddenFlag(),
				},
			})
		}
	}
}

// optionText reads one dropdown option without disturbing the selection.
func optionText(widget *tview.DropDown, index int) string {
	current, currentText := widget.GetCurrentOption()
	if index == current {
		return currentText
	}
	// tview exposes no per-index accessor, so options other than the current
	// one are published positionally.
	return "option " + strconv.Itoa(index+1)
}

// frameworkTypeFor names the widget's own Go type, which the protocol requires
// whenever the role is generic: an unrecognised widget must at least say what
// tview (or the application) called it. `%T` yields e.g. "*tview.Box" or the
// application's own "*main.Sparkline", which is exactly the distinguishing
// information a reader of an unknown node needs.
func frameworkTypeFor(primitive tview.Primitive, role protocol.Role) string {
	if role != protocol.RoleGeneric {
		return ""
	}
	return fmt.Sprintf("%T", primitive)
}

File: clients/python/src/termwright/tree.py
"""Semantic tree DTOs.

These mirror ``@termwright/protocol``'s ``tree.ts``. ``to_wire`` drops unset
optionals, because the wire schema is strict: an explicit ``null`` is a
validation failure, not "absent".
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Dict, List, Mapping, Optional, Sequence, Union


@dataclass(frozen=True)
class Rect:
    """Zero-based viewport cell coordinates."""

    row: int
    column: int
    width: int
    height: int

    def to_wire(self) -> Dict[str, int]:
        return {"row": self.row, "column": self.column, "width": self.width, "height": self.height}


@dataclass(frozen=True)
class EvidenceProvenance:
    """Origin and strength of an authoritative or diagnostic observation."""

    source: str
    method: str
    strength: str
    providerId: str

    def to_wire(self) -> Dict[str, str]:
        return {
            "source": self.source,
            "method": self.method,
            "strength": self.strength,
            "providerId": self.providerId,
        }


def framework_evidence(provider_id: str) -> EvidenceProvenance:
    """Build authoritative native framework provenance for a provider."""

    return EvidenceProvenance("framework", "native", "authoritative", provider_id)


@dataclass(frozen=True)
class Observation:
    """Evidence-qualified wire fact.

    ``known`` carries the value and provenance; authoritative ``absent`` also
    carries provenance proving that no value exists.
    """

    status: str
    value: Any = None
    evidence: Optional[EvidenceProvenance] = None
    reason: Optional[str] = None
    capability: Optional[str] = None

    def to_wire(self) -> Dict[str, Any]:
        wire: Dict[str, Any] = {"status": self.status}
        if self.status == "known":
            wire["value"] = self.value.to_wire() if hasattr(self.value, "to_wire") else self.value
            wire["evidence"] = self.evidence.to_wire() if self.evidence is not None else None
        elif self.status == "absent":
            wire["reason"] = self.reason
            wire["evidence"] = self.evidence.to_wire() if self.evidence is not None else None
        elif self.status == "unsupported":
            wire["capability"] = self.capability
            wire["reason"] = self.reason
        else:
            wire["reason"] = self.reason
        return wire


@dataclass(frozen=True)
class NodeGeometryObservations:
    displayed: Observation
    intendedRect: Observation
    visibleRect: Observation

    def to_wire(self) -> Dict[str, Any]:
        return {"displayed": self.displayed.to_wire(), "intendedRect": self.intendedRect.to_wire(), "visibleRect": self.visibleRect.to_wire()}


@dataclass(frozen=True)
class SemanticState:
    """Closed state set; unset members are omitted from the wire form."""

    disabled: Optional[bool] = None
    focused: Optional[bool] = None
    selected: Optional[bool] = None
    checked: Optional[Union[bool, str]] = None
    expanded: Optional[bool] = None
    modal: Optional[bool] = None
    busy: Optional[bool] = None
    hidden: Optional[bool] = None
    #: Every cell is outside the visible area — scrolled away, not undisplayed.
    #: Implies ``hidden``; the pair without it is refused by validation.
    offscreen: Optional[bool] = None
    readonly: Optional[bool] = None
    multiline: Optional[bool] = None
    required: Optional[bool] = None
    multiselectable: Optional[bool] = None
    orientation: Optional[str] = None
    level: Optional[int] = None
    positionInSet: Optional[int] = None
    setSize: Optional[int] = None

    def to_wire(self) -> Dict[str, Any]:
        return {
            name: value
            for name, value in self.__dict__.items()
            if value is not None
        }


@dataclass(frozen=True)
class SemanticTextRange:
    """Maps grapheme offsets of a node's text onto cell coordinates."""

    startOffset: int
    endOffset: int
    rect: Rect

    def to_wire(self) -> Dict[str, Any]:
        return {
            "startOffset": self.startOffset,
            "endOffset": self.endOffset,
            "rect": self.rect.to_wire(),
        }


@dataclass(frozen=True)
class SemanticScrollState:
    """Production application viewport state, distinct from terminal scrollback."""

    axis: str
    offset: int
    viewport: int
    extent: int

    def to_wire(self) -> Dict[str, Any]:
        return {
            "axis": self.axis,
            "offset": self.offset,
            "viewport": self.viewport,
            "extent": self.extent,
        }


@dataclass(frozen=True)
class SemanticPaintedRegion:
    """Exact viewport cells painted by one semantic recipient."""

    regionBounds: Rect
    spans: Sequence[Mapping[str, int]]

    def to_wire(self) -> Dict[str, Any]:
        return {
            "regionBounds": self.regionBounds.to_wire(),
            "spans": [dict(span) for span in self.spans],
        }


@dataclass(frozen=True)
class SemanticValueObservation:
    """A semantic value with support, absence and confidentiality preserved."""

    status: str
    value: Optional[str] = None
    sensitivity: Optional[str] = None
    evidence: Optional[EvidenceProvenance] = None
    reason: Optional[str] = None
    capability: Optional[str] = None

    def to_wire(self) -> Dict[str, Any]:
        wire: Dict[str, Any] = {"status": self.status}
        if self.value is not None:
            wire["value"] = self.value
        if self.sensitivity is not None:
            wire["sensitivity"] = self.sensitivity
        if self.evidence is not None:
            wire["evidence"] = self.evidence.to_wire()
        if self.reason is not None:
            wire["reason"] = self.reason
        if self.capability is not None:
            wire["capability"] = self.capability
        return wire


@dataclass(frozen=True)
class SemanticNode:
    """One accessible node with evidence-qualified geometry."""

    id: str
    role: str
    geometry: NodeGeometryObservations
    name: str = ""
    parentId: Optional[str] = None
    description: Optional[str] = None
    value: Optional[SemanticValueObservation] = None
    state: Optional[SemanticState] = None
    #: Application-defined JSON state. Portable flags stay in ``state``.
    extended: Optional[Mapping[str, Any]] = None
    actions: Optional[Sequence[str]] = None
    inputRecipes: Optional[Sequence[Mapping[str, Any]]] = None
    labelledBy: Optional[Sequence[str]] = None
    describedBy: Optional[Sequence[str]] = None
    textRanges: Optional[Sequence[SemanticTextRange]] = None
    testId: Optional[str] = None
    #: What the UI framework calls this widget. Required when ``role`` is
    #: ``generic``: an unrecognised widget must at least name its own type, so
    #: a reader can tell one unknown thing from another.
    frameworkType: Optional[str] = None
    #: True when this node may own children the framework cannot enumerate.
    opaqueChildren: bool = False
    #: Where this node's facts came from, as a whole.
    p: Optional[str] = None
    #: Where individual fields came from, when they differ from ``p``.
    px: Optional[Mapping[str, str]] = None
    scroll: Optional[Observation] = None
    paintedRegion: Optional[Observation] = None

    def to_wire(self) -> Dict[str, Any]:
        wire: Dict[str, Any] = {"id": self.id, "role": self.role, "name": self.name}
        if self.parentId is not None:
            wire["parentId"] = self.parentId
        if self.description is not None:
            wire["description"] = self.description
        if self.value is not None:
            wire["value"] = self.value.to_wire()
        if self.state is not None:
            state = self.state.to_wire()
            if state:
                wire["state"] = state
        if self.extended is not None:
            wire["extended"] = _canonical_extended(self.extended)
        if self.actions is not None:
            wire["actions"] = list(self.actions)
        if self.inputRecipes is not None:
            wire["inputRecipes"] = [dict(recipe) for recipe in self.inputRecipes]
        if self.labelledBy is not None:
            wire["labelledBy"] = list(self.labelledBy)
        if self.describedBy is not None:
            wire["describedBy"] = list(self.describedBy)
        if self.textRanges is not None:
            wire["textRanges"] = [item.to_wire() for item in self.textRanges]
        if self.testId is not None:
            wire["testId"] = self.testId
        if self.frameworkType is not None:
            wire["frameworkType"] = self.frameworkType
        if self.opaqueChildren:
            wire["opaqueChildren"] = True
        if self.p is not None:
            wire["p"] = self.p
        if self.px:
            wire["px"] = dict(self.px)
        wire["geometry"] = self.geometry.to_wire()
        if self.scroll is not None:
            wire["scroll"] = self.scroll.to_wire()
        if self.paintedRegion is not None:
            wire["paintedRegion"] = self.paintedRegion.to_wire()
        return wire


def _canonical_extended(value: Any) -> Any:
    """Copy JSON-like domain state with stable object-key ordering."""
    if isinstance(value, Mapping):
        return {key: _canonical_extended(value[key]) for key in sorted(value)}
    if isinstance(value, (list, tuple)):
        return [_canonical_extended(item) for item in value]
    return value


@dataclass(frozen=True)
class CursorInfo:
    """Terminal cursor position, in viewport cells."""

    row: int
    column: int
    visible: bool
    shape: Optional[str] = None

    def to_wire(self) -> Dict[str, Any]:
        wire: Dict[str, Any] = {"row": self.row, "column": self.column, "visible": self.visible}
        if self.shape is not None:
            wire["shape"] = self.shape
        return wire


@dataclass(frozen=True)
class SemanticSnapshot:
    """A whole tree for one committed render."""

    sessionId: str
    revision: int
    columns: int
    rows: int
    coordinateSpace: Observation
    hitGrid: Observation
    rootIds: Sequence[str] = field(default_factory=list)
    nodes: Sequence[SemanticNode] = field(default_factory=list)
    cursor: Optional[CursorInfo] = None
    v: int = 3

    def to_wire(self) -> Dict[str, Any]:
        wire: Dict[str, Any] = {
            "v": self.v,
            "sessionId": self.sessionId,
            "revision": self.revision,
            "columns": self.columns,
            "rows": self.rows,
        }
        if self.cursor is not None:
            wire["cursor"] = self.cursor.to_wire()
        wire["rootIds"] = list(self.rootIds)
        wire["nodes"] = [node.to_wire() for node in self.nodes]
        wire["coordinateSpace"] = self.coordinateSpace.to_wire()
        wire["hitGrid"] = self.hitGrid.to_wire()
        return wire


def snapshot_from_wire(value: Dict[str, Any]) -> SemanticSnapshot:
    """Rebuild a snapshot from an already-validated wire object."""
    def observation(raw: Mapping[str, Any], *, rect: bool = False) -> Observation:
        observed = raw.get("value")
        if rect and isinstance(observed, Mapping):
            observed = Rect(**observed)
        return Observation(
            status=raw["status"],
            value=observed,
            evidence=(EvidenceProvenance(**raw["evidence"]) if isinstance(raw.get("evidence"), dict) else None),
            reason=raw.get("reason"),
            capability=raw.get("capability"),
        )

    nodes: List[SemanticNode] = []
    for raw in value["nodes"]:
        state = raw.get("state")
        ranges = raw.get("textRanges")
        geometry = raw["geometry"]
        nodes.append(
            SemanticNode(
                id=raw["id"],
                role=raw["role"],
                name=raw.get("name", ""),
                parentId=raw.get("parentId"),
                description=raw.get("description"),
                value=(
                    SemanticValueObservation(
                        status=raw["value"]["status"],
                        value=raw["value"].get("value"),
                        sensitivity=raw["value"].get("sensitivity"),
                        evidence=(EvidenceProvenance(**raw["value"]["evidence"]) if isinstance(raw["value"].get("evidence"), dict) else None),
                        reason=raw["value"].get("reason"),
                        capability=raw["value"].get("capability"),
                    )
                    if isinstance(raw.get("value"), dict)
                    else None
                ),
                state=SemanticState(**state) if state is not None else None,
                extended=raw.get("extended"),
                actions=tuple(raw["actions"]) if raw.get("actions") is not None else None,
                labelledBy=tuple(raw["labelledBy"]) if raw.get("labelledBy") is not None else None,
                describedBy=tuple(raw["describedBy"]) if raw.get("describedBy") is not None else None,
                textRanges=tuple(
                    SemanticTextRange(
                        startOffset=item["startOffset"],
                        endOffset=item["endOffset"],
                        rect=Rect(**item["rect"]),
                    )
                    for item in ranges
                )
                if ranges is not None
                else None,
                testId=raw.get("testId"),
                frameworkType=raw.get("frameworkType"),
                opaqueChildren=raw.get("opaqueChildren", False),
                p=raw.get("p"),
                px=raw.get("px"),
                geometry=NodeGeometryObservations(
                    displayed=observation(geometry["displayed"]),
                    intendedRect=observation(geometry["intendedRect"], rect=True),
                    visibleRect=observation(geometry["visibleRect"], rect=True),
                ),
            )
        )
    cursor = value.get("cursor")
    return SemanticSnapshot(
        sessionId=value["sessionId"],
        revision=value["revision"],
        columns=value["columns"],
        rows=value["rows"],
        rootIds=tuple(value["rootIds"]),
        nodes=tuple(nodes),
        cursor=CursorInfo(**cursor) if cursor is not None else None,
        v=value["v"],
        coordinateSpace=observation(value["coordinateSpace"]),
        hitGrid=observation(value["hitGrid"]),
    )
